tpp-app/src/services/cycle/CycleService.js

```diff
@@ -0,0 +1,290 @@
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import {FLOW_LEVEL} from '../utils/constants';
+import {initializeEmptyYear, getDateString, getCalendarByYear, getSymptomsFromCalendar, getDaysDiffInclusive}  from '../utils/helpers';
+import {Symptoms} from '../utils/models';
+import differenceInCalendarDays from 'date-fns/differenceInDays';
+import isSameDay from 'date-fns/isSameDay';
+import subDays from 'date-fns/subDays';
+import addDays from 'date-fns/addDays';
+import Keys from '../utils/keys';;
+
+
+
+
+
+/**
+ * Gets the next period end date for a given date. This is not a prediction.
+ * @param {Date} searchFrom The date relative to which to find the next period end date. Must be an earlier date than today
+ * @param {Object} calendar The object containing the symptoms for this year, last year, and next year. Optional.
+ * @return {Promise} Resolves into Date object that is the closest day after searchFrom that is the end of a period
+ */
+async function getNextPeriodEnd(searchFrom, calendar = null){
+    var current = searchFrom;
+    var yesterday = subDays(current, 1);
+    var twoDaysEarlier = subDays(current, 2);
+
+    if (!calendar){
+      calendar = await getCalendarByYear(current.getFullYear());
+    }
+
+    let dateSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth()+1, current.getFullYear());
+    let yesterdaySymptoms = new Symptoms();
+    let twoDaysEarlierSymptoms = new Symptoms()
+
+    var noFlowToday = (dateSymptoms.flow === FLOW_LEVEL.NONE || dateSymptoms.flow === null) ;
+    var noFlowYesterday = (yesterdaySymptoms.flow === FLOW_LEVEL.NONE || yesterdaySymptoms.flow === null) ;
+    var flowTwoDaysEarlier = (twoDaysEarlierSymptoms.flow !== null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE)
+
+
+    while(!(noFlowToday && noFlowYesterday && flowTwoDaysEarlier)){
+
+      var tomorrow = addDays(current, 1);
+      twoDaysEarlier = yesterday;
+      yesterday = current;
+      current = tomorrow;
+
+
+      twoDaysEarlierSymptoms = yesterdaySymptoms;
+      yesterdaySymptoms = dateSymptoms;
+      dateSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth() + 1, current.getFullYear());
+      noFlowToday = (dateSymptoms.flow === FLOW_LEVEL.NONE || dateSymptoms.flow === null) ;
+      noFlowYesterday = (yesterdaySymptoms.flow === FLOW_LEVEL.NONE || yesterdaySymptoms.flow === null) ;
+      flowTwoDaysEarlier = (twoDaysEarlierSymptoms.flow !== null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE)
+    }
+
+
+
+    //return twoDaysEarlier since pattern searching for is _ _ X where X is the period
```

Comment:
this is the same comment as line 104. Is this one supposed to be X _ _ ?

Reply:
yes i copy pasted this is my b